crates/wasm-mutate/src/lib.rs

```diff
@@ -367,15 +372,27 @@ impl WasmMutate {
             RemoveExportMutator,
             SnipMutator,
             SetFunction2Unreachable,
+            PeepholeMutator,
         };
 
         let mut rnd = SmallRng::seed_from_u64(self.seed);
-        let mutator = mutators
+        let mut mutator = mutators
             .choose(&mut rnd)
             .ok_or(Error::NoMutationsAplicable)?;
 
-        let module = mutator.mutate(&self, &mut rnd, &mut info);
+        let moduleresult = mutator.mutate(&self, &mut rnd, &mut info);
 
-        Ok(module?.finish())
+        match moduleresult {
+            Ok(module) => Ok(module.finish()),
+            Err(e) => {
+                match e {
+                    Error::NotMatchingPeepholes => {
+                        // TODO, Select another mutator
+                        return Err(e);
+                    }
+                    _ => return Err(e),
```

Comment:
Nice solution :)

Reply:
However, I think that shuffling  the whole collection of mutators cannot be the best solution in terms of performance. What about having a maximum number of random rotations to select the first available mutator?

Reply:
I think we do want to try every mutation once, if every single thing before it is failing, because the alternative is that we return an error, our fuzz mutator integration code does *something*, and then libFuzzer just calls us again with a new random seed. And getting called again with a new random seed means we have to recreate the AST and reparse and all that. So we might as well try and squeeze everything out of the work we've already done before giving up.

But you're right that we don't need to shuffle the whole vec up front, we can do it incrementally and save a handful of calls into the RNG like this:

```rust
while !mutators.is_empty() {
    let i = rng.gen_range(0, mutators.len());
    let mutator = mutators.swap_remove(i);
    if let Ok(module) = mutator.mutate(...) {
        return module;
    }
}

return Err(Error::NoApplicableMutations);
```

Reply:
Gotcha

Reply:
WDYT about having this in the next iteration?